### PR TITLE
Jbrowse2

### PR DESCRIPTION
**Pull request by fubar2**
FOR CONTRIBUTOR:
* [X] - I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [X] - License permits unrestricted use (educational + commercial)
* [X] - This PR adds a new tool or tool collection

This is a very stripped-down working prototype, repurposing @hexylena's JBrowse1-in-Galaxy for JBrowse2. 

Available for testing from the [toolshed](https://toolshed.g2.bx.psu.edu/view/fubar/jbrowse2dev/af3fba6c21aa)

Most things are working in testing - screenshot shows bed bam vcf maf bigwig gff sample data tracks
image

and HiC
image

- JB1 is EOL.
- JB2 configuration json is wildly different from JB1. 
- JB2 defaults are sensible enough for many non-complicated situations.
  - for many workflows like the [TreeValGal](https://github.com/fubar2/treeval_gal) project, this will work very well as is
  - some challenges using jbrowse1 with very dense tracks in huge VGP data. 
- Only the easy part is working for the basic tracks, recycling Helena's core code, and adding hic.
- URLs are used wherever easy to do to minimise the size of JBrowse2 files folder so these are much smaller than jb1 where all were copied. Not practicable for huge files like hic. JBrowse2 lets galaxy offload existing data and sequence files to nginx range downloads.
- A _lot of work_ remains, to figure out and implement all the new track configuration options in the UI. 
   - Could not find any human readable documentation on these - need to use the browser to find track setting minutae. 
   - Nor was there anything on changing to different codon schemes that I could find. So that's a big loss - but it might be buried somewhere. The new architecture is plugins to do stuff so might need one for codons.
  - SPARQL, REST remain to be implemented.

### CODE REVIEW

File: tools/jbrowse/.shed.yml [hunk 1]
```diff
@@ -1,12 +1,12 @@
 categories:
 - Sequence Analysis
-description: JBrowse Genome Browser integrated as a Galaxy Tool
+description: JBrowse2 Genome Browser integrated as a Galaxy Tool
```
Comment (bgruening) [inline tools/jbrowse/.shed.yml:3]: Can you please revert this change to the older jbrowse?

Comment (fubar2) [reply tools/jbrowse/.shed.yml:3]: yes - I did as soon as I saw planemo testing the wrong xml
+ GITHUB_REF=refs/pull/5695/merge
+ '[' 'tools/jbrowse
tools/jbrowse2' == '' ']'
+ echo -n 'tools/jbrowse
tools/jbrowse2'
+ echo -n tools/jbrowse2/jbrowse2.xml
+ echo -n 20
+ '[' lint == lint ']'
+ lint_fail=false
+ mapfile -t REPO_ARRAY
+ for DIR in "${REPO_ARRAY[@]}"
+ '[' false '!=' true ']'
+ planemo shed_lint --tools --ensure_metadata --urls --report_level all --fail_level warn --recursive tools/jbrowse
+ tee -a lint_report.txt
Linting repository /home/runner/work/tools-iuc/tools-iuc/tools/jbrowse

Comment (fubar2) [reply tools/jbrowse/.shed.yml:3]: oh I rewrote the wrong one too ! Sorry - yes of course.

Comment (fubar2) [reply tools/jbrowse/.shed.yml]: done @bgruening Sorry...

Comment (fubar2) [reply tools/jbrowse/.shed.yml]: Still seeing CI lint failing with the wrong xml!
```
tools/jbrowse2' == '' ']'
+ echo -n 'tools/jbrowse
tools/jbrowse2'
+ echo -n tools/jbrowse2/jbrowse2.xml
+ echo -n 20
+ '[' lint == lint ']'
+ lint_fail=false
+ mapfile -t REPO_ARRAY
+ for DIR in "${REPO_ARRAY[@]}"
+ '[' false '!=' true ']'
+ planemo shed_lint --tools --ensure_metadata --urls --report_level all --fail_level warn --recursive tools/jbrowse
+ tee -a lint_report.txt
**Linting repository /home/runner/work/tools-iuc/tools-iuc/tools/jbrowse**
Applying linter expansion... CHECK
.. INFO: Included files all found.
```

Comment (fubar2) [reply tools/jbrowse/.shed.yml]: ok...passes now. Ouch. That was annoying clumsiness.

**Comment by fubar2** [comment]
Odd. Passes planemo test here:
```

d0c-e303-426f-8b32-8b8dd5b8df42.dat to /tmp/tmp9pxo1ggt/files/2/b/e/dataset_2bea3d0c-e303-426f-8b32-8b8dd5b8df42.dat
galaxy.model.metadata DEBUG 2023-12-27 21:00:47,645 [pN:main.1,p:3223405,tN:LocalRunner.work_thread-1] loading metadata from file for: HistoryDatasetAssociation 10
galaxy.jobs DEBUG 2023-12-27 21:00:47,719 [pN:main.1,p:3223405,tN:LocalRunner.work_thread-1] job_wrapper.finish for job 10 executed (380.086 ms)

Shut down
supervisord has terminated
All 4 test(s) executed passed.
jbrowse2 (Test #1): passed
jbrowse2 (Test #2): passed
jbrowse2 (Test #3): passed
jbrowse2 (Test #4): passed
(.venv) ross@pn50:~/rossgit/tools-iuc/tools/jbrowse2$ planemo --version
planemo, version 0.75.19

```

**Comment by bgruening** [comment]
Please note that the CI is running 'planemo test --biocontainers '

**Review by fubar2 — COMMENTED** [review]
@bgruening
Thanks for fixing the version command calling jbrowse instead of jbrowse2 and adding the biocontainer!
That first problem made the tool fail silently when profile was set to 22.05 - no error message.

**Comment by fubar2** [comment]
added the JBrowse2 author maf plugin and followed the recipe.
It generates something but needs a proper test. Will look for data.

File: tools/jbrowse2/jbrowse2.xml [hunk 2]
```diff
@@ -0,0 +1,532 @@
...
+            <param name="standalone" value="minimal" />
+            <repeat name="track_groups">
+                <param name="category" value="Auto Coloured" />
+                <repeat name="data_tracks">
+                    <conditional name="data_format">
+                        <param name="data_format_select" value="pileup"/>
+                        <param name="annotation" value="bam/merlin-sample.bam"/>
+                    </conditional>
+                </repeat>
+            </repeat>
```
Comment (bgruening) [inline tools/jbrowse2/jbrowse2.xml:378]: We should use this more explicit syntax going forward. The tests will get complex over times and this one is more readable.

```suggestion
            <conditional name="reference_genome">
                 <param name="genome_type_select" value="history"/>
                 <param name="genome" value="merlin.fa"/>
            </conditional>
            <param name="standalone" value="minimal" />
            <repeat name="track_groups">
                <param name="category" value="Auto Coloured" />
                <repeat name="data_tracks">
                    <conditional name="data_format">
                        <param name="data_format_select" value="pileup"/>
                        <param name="annotation" value="bam/merlin-sample.bam"/>
                    </conditional>
                </repeat>
            </repeat>
```

**Comment by fubar2** [comment]
MAF and VCF fake data:
image

**Comment by fubar2** [comment]
@bgruening - just realised that the HiC is slow because it's a terrible idea to copy huge data into each new jbrowse files path, if it can be avoided by serving urls for each file already in Galaxy instead. It may be slower but it will keep the instance to the javascript size.
So let's hold of until this is done and tested?

**Comment by fubar2** [comment]
Reference fasta now served via URL so not needed inside the jbrowse files directory
Indexing takes ages for hg19 sized files but it does work
hic files are also served via URL
Any files already in Galaxy like bed/bam/vcf can be served the same way to minimise the size of the files that need to be moved.
@bgruening - let's go with this as is for now - plenty more work to do but probably useful enough to expose...

**Comment by fubar2** [comment]
@bgruening - why does latest planemo lint work here but throws an lxml XMLSchemaParseError error in CI?
```
Applying linter citations... CHECK
.. CHECK: Found 1 likely valid citations.
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.7.17/x64/bin/planemo", line 8, in <module>
    sys.exit(planemo())
  File "/opt/hostedtoolcache/Python/3.7.17/x64/lib/python3.7/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
  File "/opt/hostedtoolcache/Python/3.7.17/x64/lib/python3.7/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
  File "/opt/hostedtoolcache/Python/3.7.17/x64/lib/python3.7/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/opt/hostedtoolcache/Python/3.7.17/x64/lib/python3.7/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/opt/hostedtoolcache/Python/3.7.17/x64/lib/python3.7/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
  File "/opt/hostedtoolcache/Python/3.7.17/x64/lib/python3.7/site-packages/click/decorators.py", line 92, in new_func
    return ctx.invoke(f, obj, *args, **kwargs)
  File "/opt/hostedtoolcache/Python/3.7.17/x64/lib/python3.7/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
  File "/opt/hostedtoolcache/Python/3.7.17/x64/lib/python3.7/site-packages/planemo/cli.py", line 102, in handle_blended_options
    return f(*args, **kwds)
  File "/opt/hostedtoolcache/Python/3.7.17/x64/lib/python3.7/site-packages/planemo/commands/cmd_shed_lint.py", line 62, in cli
    exit_code = shed.for_each_repository(ctx, lint, paths, **kwds)
  File "/opt/hostedtoolcache/Python/3.7.17/x64/lib/python3.7/site-packages/planemo/shed/__init__.py", line 762, in for_each_repository
    ret_codes.append(function(realized_repository))
  File "/opt/hostedtoolcache/Python/3.7.17/x64/lib/python3.7/site-packages/planemo/commands/cmd_shed_lint.py", line 59, in lint
    return shed_lint.lint_repository(ctx, realized_repository, **kwds)
  File "/opt/hostedtoolcache/Python/3.7.17/x64/lib/python3.7/site-packages/planemo/shed_lint.py", line 111, in lint_repository
    tools_failed = lint_repository_tools(ctx, realized_repository, lint_ctx, lint_args)
  File "/opt/hostedtoolcache/Python/3.7.17/x64/lib/python3.7/site-packages/planemo/shed_lint.py", line 129, in lint_repository_tools
    lint_tool_source_with(lint_ctx, tool_source, extra_modules=lint_args["extra_modules"])
  File "/opt/hostedtoolcache/Python/3.7.17/x64/lib/python3.7/site-packages/galaxy/tool_util/lint.py", line 341, in lint_tool_source_with
    lint_context.lint(name, value, tool_xml)
  File "/opt/hostedtoolcache/Python/3.7.17/x64/lib/python3.7/site-packages/galaxy/tool_util/lint.py", line 184, in lint
    lint_func(lint_target, self)
  File "/opt/hostedtoolcache/Python/3.7.17/x64/lib/python3.7/site-packages/planemo/linters/xsd.py", line 17, in lint_tool_xsd
    planemo.lint.lint_xsd(lint_ctx, TOOL_XSD, tf.name)
  File "/opt/hostedtoolcache/Python/3.7.17/x64/lib/python3.7/site-packages/planemo/lint.py", line 93, in lint_xsd
    validation_result = validator.validate(schema_path, path)
  File "/opt/hostedtoolcache/Python/3.7.17/x64/lib/python3.7/site-packages/planemo/xml/validation.py", line 50, in validate
    xsd = etree.XMLSchema(xsd_doc)
  File "src/lxml/xmlschema.pxi", line 89, in lxml.etree.XMLSchema.__init__
lxml.etree.XMLSchemaParseError: complex type 'Output': The content model is not determinist., line 5329
```

**Comment by fubar2** [comment]
Help wanted - getting maf to work requires changes to blastxml_to_gapped_gff3.py to use SimpleLocation with strand (already made) and changes to bcbio-gff to move strand from the feature to feature.location after updates to biopython.
```
Traceback (most recent call last):
  File "/evol/galaxytf23.1.3/local_tools/jbrowse2/blastxml_to_gapped_gff3.py", line 268, in <module>
    GFF.write([rec], sys.stdout)
  File "/evol/galaxytf23.1.3/database/dependencies/_conda/envs/mulled-v1-95c551f13f10fa5d1c18116a0042517a772496547735852a5970f20602646882/lib/python3.10/site-packages/BCBio/GFF/GFFOutput.py", line 202, in write
    return writer.write(recs, out_handle, include_fasta)
  File "/evol/galaxytf23.1.3/database/dependencies/_conda/envs/mulled-v1-95c551f13f10fa5d1c18116a0042517a772496547735852a5970f20602646882/lib/python3.10/site-packages/BCBio/GFF/GFFOutput.py", line 84, in write
    id_handler = self._write_feature(sf, rec.id, out_handle,
  File "/evol/galaxytf23.1.3/database/dependencies/_conda/envs/mulled-v1-95c551f13f10fa5d1c18116a0042517a772496547735852a5970f20602646882/lib/python3.10/site-packages/BCBio/GFF/GFFOutput.py", line 124, in _write_feature
    if feature.strand == 1:
AttributeError: 'SeqFeature' object has no attribute 'strand'
```

**Comment by abretaud** [comment]
Nice to see progress! Have you been able to reuse some code from https://github.com/galaxyproject/tools-iuc/pull/3997 or you restarted from scratch?

**Comment by bgruening** [comment]
@fubar2 tests are passing now again. Thanks to @mvdbeek!

**Comment by fubar2** [comment]
> Nice to see progress! Have you been able to reuse some code from #3997 or you restarted from scratch?

@abretaud: Helena's JB1 code is extensively reused but unfortunately I didn't know you'd done all that before I started hacking on it...

Been around in circles a couple of times because config.json being generated in Galaxy works fine in Galaxy, but when downloaded cannot be understoood by the desktop client - there seems to be a complete disconnect unfortunately - but all the common formats work and with a PR for bcbio, maf will work too.

**Comment by fubar2** [comment]
@bgruening: after a lot of pfaffing about, it turns out that the JBrowse2 desktop client cannot understand a setup that works perfectly well in Galaxy or when served locally by nginx. 
It seems that as part of the upgrading process there is now a complete disconnect between the config.json required for the desktop and a web server. 
Very annoying - we need to talk to them urgently...
At present, JB2 works fine in Galaxy. Downloaded tarballs can be unpacked and served statically from nginx easily enough but cannot be used by the desktop client.
Advice appreciated :)

**Comment by fubar2** [comment]
@bgruening: the python code for a [simple web server](https://gist.github.com/glowinthedark/b99900abe935e4ab4857314d647a9068) serves an unpacked archive from this tool.
It won't open with a web browser but now it's added to every output archive, so after unarchiving and changing to the root, they must run:

`python servjb2.py`
It has optional parameters:` --root [path to the unpacked archive. default cwd] --port [default 8080] --bind [default 0.0.0.0]`

Best I can think of - it may be some time before there is harmonisation between jbrowse2 desktop and web.

**Comment by fubar2** [comment]
@bgruening: Stumped again. Passes without biocontainers but fails using biocontainers.
>Command '['jbrowse', 'create', '-f', '/tmp/tmpa3pi9_9n/job_working_directory/000/2/outputs/dataset_755c5b88-c40d-429c-a1bf-fad6dbe1b72c_files']' returned non-zero exit status 1.
So it's not allowed to write to the jbrowse2 output folder - do I need to fix anything?

**Comment by fubar2** [comment]
Closed -  #5697 merges @abretaud's PR content.
